operator/v1/types_network.go

```diff
@@ -53,6 +53,7 @@ type NetworkList struct {
 
 // NetworkSpec is the top-level network configuration object.
 // +kubebuilder:validation:XValidation:rule="!has(self.defaultNetwork) || !has(self.defaultNetwork.ovnKubernetesConfig) || !has(self.defaultNetwork.ovnKubernetesConfig.gatewayConfig) || !has(self.defaultNetwork.ovnKubernetesConfig.gatewayConfig.ipForwarding) || self.defaultNetwork.ovnKubernetesConfig.gatewayConfig.ipForwarding == oldSelf.defaultNetwork.ovnKubernetesConfig.gatewayConfig.ipForwarding || self.defaultNetwork.ovnKubernetesConfig.gatewayConfig.ipForwarding == 'Restricted' || self.defaultNetwork.ovnKubernetesConfig.gatewayConfig.ipForwarding == 'Global'",message="invalid value for IPForwarding, valid values are 'Restricted' or 'Global'"
+// +openshift:validation:FeatureGateAwareXValidation:featureGate=BGP,rule="self.advancedRouting == 'Enabled' || !has(self.defaultNetwork.ovnKubernetesConfig) || self.defaultNetwork.ovnKubernetesConfig.routeAdvertisements != 'Enabled'",message="RouteAdvertisements cannot be Enabled if AdvancedRouting is not Enabled"
```

Comment:
Why is the advanced routing feature at a different level tan the routeAdvertisements feature? If they are tied together I would expect something like a discriminated union that shows they are explicitly linked, right now they are "far" away from each other so this validation will likely not be intuitive to end users

Reply:
This answer might also help answer https://github.com/openshift/api/pull/1944#discussion_r1665480966 and https://github.com/openshift/api/pull/1944#discussion_r1665483206

Advance routing in practice will deploy FFR which is an advanced routing protocol suite. This enables a lot of potential use cases, not all of which involve the pod network (ovn-kubernetes). Right now there is two:

- Supporting BGP in ovn-kubernetes
- As the FFR backend for MetalLB operator, does not involve ovn-kubernetes at all

Of course, the two use cases at the same time are also possible.

So the proposal here is to keep both things separated. 

Yes, a tri-state would work right now (`EnabledWithRouteAdvertisement`, `EnabledWithoutRouteAdvertisement`, Disabled) but the potential for expansion of just FFR configuration options in the future is big in a way that having its own section completely independent and un-mixed from the network plugins would just make sense and be a benefit.

Actually right now thinking on making AdvancedRouting a struct to hold other future options
```
type AdvancedRouting struct {
    // Enabled/Disabled
    BGP string
}
```

Yes, there is configuration dependent of each other that might feel "far" away but I think it just corresponds to the nature of things. We can help it with proper validation and doc.

Reply:
I would be tempted to include, as you've proposed, an advanced routing struct that has some enabled/disabled state within it, as you've done, and then make that a discriminated union with a struct underneath with optional fields within.

If we go that route, we could have various compatible options join this lower struct in the future, and, in theory, the pod networking side could be in there as well.

With the pod networking side of this, is there anything in the current location that the field depends on, eg a higher level union?

I was thinking along the lines of

```
advancedRouting:
  type: BGP
  bgp:
    ffr:
      ...
    ovnKubernetesConfig:
      routeAdvertisement: Enabled
    < insert other fields >
```

But I'm wondering if that also then depends on checking that we are using OVN. I would probably lean to putting this in the lesser commonly used place, so that those who are using say advancedRouting without OVN see the field but cannot use it, rather than those using OVN without advancedRouting, since I suspect the former is a smaller set of customers.

Bit of an awkward relation this one

Reply:
> I would be tempted to include, as you've proposed, an advanced routing struct that has some enabled/disabled state within it, as you've done, and then make that a discriminated union with a struct underneath with optional fields within.
> 
> If we go that route, we could have various compatible options join this lower struct in the future, and, in theory, the pod networking side could be in there as well.
> 
> With the pod networking side of this, is there anything in the current location that the field depends on, eg a higher level union?
> 
> I was thinking along the lines of
> 
> ```
> advancedRouting:
>   type: BGP
>   bgp:
>     ffr:
>       ...
>     ovnKubernetesConfig:
>       routeAdvertisement: Enabled
>     < insert other fields >
> ```
> 
> But I'm wondering if that also then depends on checking that we are using OVN. I would probably lean to putting this in the lesser commonly used place, so that those who are using say advancedRouting without OVN see the field but cannot use it, rather than those using OVN without advancedRouting, since I suspect the former is a smaller set of customers.
> 
> Bit of an awkward relation this one

I think there is no way to avoid some level of disjoint-ness / awkwardness here. The awkwardness with your approach is that we already have a place for ovn-kubernetes related settings (user might be already familiar with), and that is exactly where routeAdvertisement is defined in the current proposal; and it solves the matter of checking if we are using OVN because otherwise the current containing struct wouldn't be defined. So we either keep ovn-kubernetes settings together, or we keep routing settings together but we can't have both. If you think your approach is more natural we can go with it. I will ask other opinions though.

Looking forward, with FRR (sorry, misspelled this one before), there is different protocols that could be enabled: BGP, OSPF, RIP, ...
Each of them may be enabled at the same time, and may or may not have extra configuration parameters (we don't have any specific right now), aside from global routing config parameters. so it would then look more like
```
advancedRouting:
  protocols:
     - BGP
  # potentially if we had any BGP parameters, I guess validation could be done against the elements of the list above
  bgp:
    # potentially if we wanted to to have ovn-k config here
    ovnKubernetesConfig:
      routeAdvertisement: Enabled
```

Reply:
> So we either keep ovn-kubernetes settings together, or we keep routing settings together but we can't have both.

An advantage of the approach I suggested is that you can have one check (bgp ovn only allowed if default network ovn is present) and have many fields within the bgp ovn config, where with the alternate approach, you'd have potentially many future checks if you added more bgp related options directly into the default network ovn kube config.

> Each of them may be enabled at the same time, and may or may not have extra configuration parameters (we don't have any specific right now), aside from global routing config parameters. so it would then look more like

Well that's a fun kind of DU 😅 I'm sure we can come up with something like that.

Reply:
> > So we either keep ovn-kubernetes settings together, or we keep routing settings together but we can't have both.
> 
> An advantage of the approach I suggested is that you can have one check (bgp ovn only allowed if default network ovn is present) and have many fields within the bgp ovn config, where with the alternate approach, you'd have potentially many future checks if you added more bgp related options directly into the default network ovn kube config.

True although we deal with the checks and not the user. We don't plan for or expect more parameters in the OVN side (as much as we expect extra parameters for advanced routing on the other side) so current proposal also is based on that. I hear your suggestion however, let me get back to you after I discuss with the team.